### PR TITLE
migrate tslint to eslint, and typescript  4.2.4->4.9.5

### PR DESCRIPTION
**Pull request by blueww**
Thanks for contribution! Please go through following checklist before sending PR.

[Roadmap: TSLint -> ESLint · Issue #4534 · palantir/tslint (github.com)](https://github.com/palantir/tslint/issues/4534)

[TSLint is deprecated: how to upgrade to ESlint - Darragh ORiordan](https://www.darraghoriordan.com/2020/03/06/upgrade-tslint-deprecated-to-eslint/)

Also move the pipeline on Node.js 10/12 to Node.js 14/16/18, since the upgrade code always fail on node 10.
Before: 
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/weiwei/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/weiwei/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Scenario/Node |   | 10 | 12 | 14 | 16 | 18
-- | -- | -- | -- | -- | -- | --
Blob | ubuntu-latest | y | y |   |   |  
  | ubuntu-latest | y | y |   |   |  
  | windows-latest | y | y |   |   |  
  | macOS-latest | y | y |   |   |  
Blob-SQL | ubuntu-latest | y | y |   |   |  
Queue | ubuntu-latest | y | y |   |   |  
  | windows-latest | y | y |   |   |  
  | macOS-latest | y | y |   |   |  
Table | ubuntu-latest |   |   | y | y |  
  | windows-latest |   |   | y | y |  
  | macOS-latest |   |   | y | y |  
VSC | ubuntu-latest |   |   |   | y |  
  | windows-latest |   |   | y | y |  
  | macOS-latest |   |   | y | y |  
docker | ubuntu-latest |   |   |   |   |  
governance | ubuntu-latest |   |   |   |   |  
exe | ubuntu-latest |   |   | y |   |  
binary Test   Linux | ubuntu-latest |   | y |   |   |  

</body>

</html>

After the PR: 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/weiwei/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/weiwei/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Scenario/Node |   | 10 | 12 | 14 | 16 | 18
-- | -- | -- | -- | -- | -- | --
Blob | ubuntu-20.04 |   | y | y |    |   
  | ubuntu-latest |   |   |   | y | y
  | windows-latest |   |   | y | y |  
  | macOS-latest |   |   | y | y |  
Blob-SQL | ubuntu-latest |   |   | y | y |  
Queue | ubuntu-latest |   |   | y | y |  
  | windows-latest |   |   |   | y | y
  | macOS-latest |   |   | y | y |  
Table | ubuntu-latest |   |   | y | y |  
  | windows-latest |   |   | y | y |  
  | macOS-latest |   |   |   | y | y
VSC | ubuntu-latest |   |   |   | y |  
  | windows-latest |   |   | y | y |  
  | macOS-latest |   |   | y | y |  
docker | ubuntu-latest |   |   |   |   |  
governance | ubuntu-latest |   |   |   |   |  
exe | ubuntu-latest |   |   | y |   |  
binary Test   Linux | ubuntu-latest |   | y |   |   |  

</body>

</html>

### PR Branch Destination

- For Azurite V3, please send PR to `main` branch.
- For legacy Azurite V2, please send PR to `legacy-dev` branch.

### Always Add Test Cases

Make sure test cases are added to cover the code change.

### Add Change Log

Add change log for the code change in `Upcoming Release` section in `ChangeLog.md`.

### Development Guideline

Please go to CONTRIBUTION.md for steps about setting up development environment and recommended Visual Studio Code extensions.

### CODE REVIEW

**Comment by blueww** [comment]
It looks blob test can't finish on node 14/16 with Linux/Mac. (node 12, or Windows are good.)
Actually, all test cases has passed, but the `npm run test:blob` never finish, so the pipeline timeout.

@XiaoningLiu , @edwin-huber , do you have any idea how to fix it?

**Comment by edwin-huber** [comment]
We can compare it to a previous blob test run and then check which tests are not finishing.

**Comment by edwin-huber** [comment]
@blueww it looks like it is "Blob OAuth Basic" tests and the "Create container with not exist Account, return 404", which seems to have been causing issues for a number of PRs before this.

**Comment by blueww** [comment]
@edwin-huber 

Actually all test cases success (can see the test log for all cases), but the "npm run test:blob" command never exist, so timeout.

It looks related with the last 3 cases in https://github.com/Azure/Azurite/blob/main/tests/blob/blockblob.highlevel.test.ts.

**Comment by blueww** [comment]
@edwin-huber , @XiaoningLiu 

It looks following 2 cases cause the timeout of blob test on node 14/16/18 on Linux/Mac.
Do you have any idea how to fix them?

- "blobclient.download should download full data successfully when internal stream unexpected ends @loki @sql"
- "blobclient.download should download data failed when exceeding max stream retry requests @loki @sql"

**Comment by blueww** [comment]
I have refer https://github.com/Azure/Azurite/pull/1755, and fixed the test timeout issue.
Would you please help to review this PR, then we can merge it. (Espicially the changes in 'LokiQueueMetadataStore.ts", typescript 4.9.5 will always report error if not do the change. Do you have a better idea to change it?)

**Review by edwin-huber — APPROVED** [review]
I am OK with this initial move to eslint.
We should then stabilize the dev tools on the new version, especially for prettier and code style, I would prefer a replacement or to add it back in.

File: package.json [hunk 1]
```diff
@@ -69,22 +69,22 @@
...
     "azure-storage": "^2.10.3",
     "cross-env": "^6.0.3",
     "cross-var": "^1.1.0",
+    "eslint": "^8.35.0",
     "find-process": "^1.4.4",
     "husky": "^8.0.1",
     "lint-staged": "^13.0.0",
     "mocha": "^5.2.0",
     "pkg": "^5.3.0",
-    "prettier": "^2.2.1",
```
[the remark this answers is not in the document]
Comment (blueww) [reply package.json:81]: Good catch.
Add back.

File: package.json [hunk 2]
```diff
@@ -283,12 +283,12 @@
...
-    "test:blob:sql": "npm run lint && cross-env cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 AZURITE_TEST_DB=mysql://root:my-secret-pw@127.0.0.1:3306/azurite_blob_test mocha --compilers ts-node/register --no-timeouts --grep @sql --recursive tests/blob/*.test.ts tests/blob/**/*.test.ts",
-    "test:blob:sql:ci": "npm run lint && cross-env cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 AZURITE_TEST_DB=mysql://root:my-secret-pw@127.0.0.1:13306/azurite_blob_test mocha --compilers ts-node/register --no-timeouts --grep @sql --recursive tests/blob/*.test.ts tests/blob/**/*.test.ts",
-    "test:queue": "npm run lint && cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 mocha --compilers ts-node/register --no-timeouts --recursive tests/queue/*.test.ts tests/queue/**/*.test.ts",
+    "lint": "npx eslint src/**/*.ts",
+    "test": "npm run lint && cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 mocha --compilers ts-node/register --no-timeouts --grep @loki --recursive --exit tests/**/*.test.ts tests/**/**/*.test.ts",
+    "test:blob": "npm run lint && cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 mocha --compilers ts-node/register --no-timeouts --grep @loki --recursive --exit tests/blob/*.test.ts tests/blob/**/*.test.ts",
```
[the remark this answers is not in the document]
Comment (blueww) [reply package.json:289]: @edwin-huber 

The test run are blocked on following 2 cases, do you have any idea how to fix them?

- "blobclient.download should download full data successfully when internal stream unexpected ends @loki https://github.com/Sql"
- "blobclient.download should download data failed when exceeding max stream retry requests @loki https://github.com/Sql"

If can't have a quick fix for them, how about we open an issue to track the 2 cases, and add "--exit" to unblock eslint upgrade?

File: .eslintrc.js
Comment (XiaoningLiu) [inline .eslintrc.js:1]: Can we open an item/issue tracking the disabled lint rules? We can bring them back like on Table GA.

File: package.json.bak
Comment (XiaoningLiu) [inline package.json.bak:1]: What does this file do?

Comment (blueww) [reply package.json.bak:1]: Removed.

File: .eslintrc.js
Comment (blueww) [reply .eslintrc.js:1]: task opened: https://github.com/Azure/Azurite/issues/1833

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (blueww) [reply package.json:289]: Track issue is opened: https://github.com/Azure/Azurite/issues/1834

**Review by XiaoningLiu — APPROVED** [review]

**Comment by blueww** [comment]
with https://github.com/Azure/Azurite/pull/2028 merge,  (teston node 12 changed to node 20), the pipeline changed to :

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/weiwei/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/weiwei/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



Scenario/Node |   | 10 | 12 | 14 | 16 | 18 | 20
-- | -- | -- | -- | -- | -- | -- | --
Blob | ubuntu-20.04 |   |   | y |    |    | y|     
  | ubuntu-latest |   |   |   | y | y
  | windows-latest |   |   | y | y |  
  | macOS-latest |   |   | y | y |  
Blob-SQL | ubuntu-latest |   |   | y | y |  
Queue | ubuntu-latest |   |   | y | y |  
  | windows-latest |   |   |   | y | y
  | macOS-latest |   |   | y | y |  
Table | ubuntu-latest |   |   | y | y |  
  | windows-latest |   |   | y | y |  
  | macOS-latest |   |   |   | y | y
VSC | ubuntu-latest |   |   |   | y |  
  | windows-latest |   |   | y | y |  
  | macOS-latest |   |   | y | y |  
docker | ubuntu-latest |   |   |   |   |  
governance | ubuntu-latest |   |   |   |   |  
exe | ubuntu-latest |   |   | y |   |  
binary Test   Linux | ubuntu-latest |   |  |   |   |  y |



</body>

</html>

**Pull request merged**